src/btree/cursor.ts

```diff
@@ -0,0 +1,81 @@
+import { BTree, ReferencedValue } from "./btree";
+import { BTreeNode } from "./node";
+
+export class BTreeCursor {
+  private readonly btree: BTree;
+
+  private uniqueEntriesPromise: Promise<number> | null = null;
+
+  constructor(btree: BTree) {
+    this.btree = btree;
+  }
+
+  async first(): Promise<ReferencedValue> {
+    let { rootNode } = await this.btree.root();
+    if (!rootNode) {
+      throw new Error("unable to get root node");
+    }
+
+    let currNode = await this.btree.readNode(rootNode.pointer(0));
+
+    while (!currNode.leaf()) {
+      const childPointer = currNode.pointer(0);
+      currNode = await this.btree.readNode(childPointer);
+    }
+
+    return currNode.keys[0];
+  }
+
+  async last(): Promise<ReferencedValue> {
+    let { rootNode } = await this.btree.root();
+    if (!rootNode) {
+      throw new Error("unable to get root node");
+    }
+
+    let currNode = await this.btree.readNode(
+      rootNode.pointer(rootNode.numPointers() - 1),
+    );
+
+    while (!currNode.leaf()) {
+      const childPointer = currNode.pointer(currNode.numPointers() - 1);
+      currNode = await this.btree.readNode(childPointer);
+    }
+
+    return currNode.keys[currNode.keys.length - 1];
+  }
+
+  // Counts the number of unique entries in the datafile
+  uniqueEntries(): Promise<number> {
+    if (!this.uniqueEntriesPromise) {
+      this.uniqueEntriesPromise = this.computeUniqueLeafs();
+    }
+
+    return this.uniqueEntriesPromise;
+  }
```

Comment:
This is needed to compute `N` in the ranking fn: 

```
N is the total number of documents on the content node.
```

In this case, this would be size of a set containing all leaf nodes' memory pointer offset